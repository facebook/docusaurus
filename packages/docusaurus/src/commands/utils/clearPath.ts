/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {rm} from 'fs/promises';
import {PerfLogger} from '@docusaurus/logger';

/**
 * @param pathToClear
 */
export default async function clearPath(pathToClear: string): Promise<void> {
  return PerfLogger.async(
    `clearPath ${path.relative(process.cwd(), pathToClear)}`,
    async () => {
      await rm(pathToClear, {recursive: true, force: true});
    },
  );
}
