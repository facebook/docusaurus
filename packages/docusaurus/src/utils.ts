/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import {AsyncLocalStorage} from 'async_hooks';
import logger from '@docusaurus/logger';

// For now this is a private env variable we use internally
// But we'll want to expose this feature officially some day
export const PerfDebuggingEnabled: boolean =
  !!process.env.DOCUSAURUS_PERF_LOGGER;

const Thresholds = {
  min: 5,
  yellow: 100,
  red: 1000,
};

const PerfPrefix = logger.yellow(`[PERF] `);

// This is what enables to "see the parent stack" for each log
// Parent1 > Parent2 > Parent3 > child trace
const ParentPrefix = new AsyncLocalStorage<string>();
function applyParentPrefix(label: string) {
  const parentPrefix = ParentPrefix.getStore();
  return parentPrefix ? `${parentPrefix} > ${label}` : label;
}

type PerfLoggerAPI = {
  start: (label: string) => void;
  end: (label: string) => void;
  log: (message: string) => void;
  async: <Result>(
    label: string,
    asyncFn: () => Result | Promise<Result>,
  ) => Promise<Result>;
};

function createPerfLogger(): PerfLoggerAPI {
  if (!PerfDebuggingEnabled) {
    const noop = () => {};
    return {
      start: noop,
      end: noop,
      log: noop,
      async: async (_label, asyncFn) => asyncFn(),
    };
  }

  const formatDuration = (duration: number): string => {
    if (duration > Thresholds.red) {
      return logger.red(`${(duration / 1000).toFixed(2)} seconds!`);
    } else if (duration > Thresholds.yellow) {
      return logger.yellow(`${duration.toFixed(2)} ms`);
    } else {
      return logger.green(`${duration.toFixed(2)} ms`);
    }
  };

  const logDuration = (label: string, duration: number) => {
    if (duration < Thresholds.min) {
      return;
    }
    console.log(`${PerfPrefix + label} - ${formatDuration(duration)}`);
  };

  const start: PerfLoggerAPI['start'] = (label) => performance.mark(label);

  const end: PerfLoggerAPI['end'] = (label) => {
    const {duration} = performance.measure(label);
    performance.clearMarks(label);
    logDuration(applyParentPrefix(label), duration);
  };

  const log: PerfLoggerAPI['log'] = (label: string) =>
    console.log(PerfPrefix + applyParentPrefix(label));

  const async: PerfLoggerAPI['async'] = async (label, asyncFn) => {
    const finalLabel = applyParentPrefix(label);
    const before = performance.now();
    const result = await ParentPrefix.run(finalLabel, () => asyncFn());
    const duration = performance.now() - before;
    logDuration(finalLabel, duration);
    return result;
  };

  return {
    start,
    end,
    log,
    async,
  };
}

export const PerfLogger: PerfLoggerAPI = createPerfLogger();
