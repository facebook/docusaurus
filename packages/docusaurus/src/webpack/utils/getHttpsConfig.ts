/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import path from 'path';
import crypto from 'crypto';
import logger from '@docusaurus/logger';

// Ensure the certificate and key provided are valid and if not
// throw an easy to debug error
function validateKeyAndCerts({
  cert,
  key,
  keyFile,
  crtFile,
}: {
  cert: Buffer;
  key: Buffer;
  keyFile: string;
  crtFile: string;
}) {
  let encrypted: Buffer;
  try {
    // publicEncrypt will throw an error with an invalid cert
    encrypted = crypto.publicEncrypt(cert, Buffer.from('test'));
  } catch (err) {
    logger.error`The certificate path=${crtFile} is invalid.`;
    throw err;
  }

  try {
    // privateDecrypt will throw an error with an invalid key
    crypto.privateDecrypt(key, encrypted);
  } catch (err) {
    logger.error`The certificate key path=${keyFile} is invalid.`;
    throw err;
  }
}

// Read file and throw an error if it doesn't exist
async function readEnvFile(file: string, type: string) {
  if (!(await fs.pathExists(file))) {
    throw new Error(
      `You specified ${type} in your env, but the file "${file}" can't be found.`,
    );
  }
  return fs.readFile(file);
}

// Get the https config
// Return cert files if provided in env, otherwise just true or false
export default async function getHttpsConfig(): Promise<
  boolean | {cert: Buffer; key: Buffer}
> {
  const appDirectory = await fs.realpath(process.cwd());
  const {SSL_CRT_FILE, SSL_KEY_FILE, HTTPS} = process.env;
  const isHttps = HTTPS === 'true';

  if (isHttps && SSL_CRT_FILE && SSL_KEY_FILE) {
    const crtFile = path.resolve(appDirectory, SSL_CRT_FILE);
    const keyFile = path.resolve(appDirectory, SSL_KEY_FILE);
    const config = {
      cert: await readEnvFile(crtFile, 'SSL_CRT_FILE'),
      key: await readEnvFile(keyFile, 'SSL_KEY_FILE'),
    };

    validateKeyAndCerts({...config, keyFile, crtFile});
    return config;
  }
  return isHttps;
}
