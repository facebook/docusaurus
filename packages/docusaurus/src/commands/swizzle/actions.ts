/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import path from 'path';
import _ from 'lodash';
import logger from '@docusaurus/logger';
import {Globby, posixPath, THEME_PATH} from '@docusaurus/utils';
import {askSwizzleAction} from './prompts';
import type {SwizzleAction, SwizzleComponentConfig} from '@docusaurus/types';
import type {SwizzleCLIOptions} from './common';

export const SwizzleActions: SwizzleAction[] = ['wrap', 'eject'];

export async function getAction(
  componentConfig: SwizzleComponentConfig,
  options: Pick<SwizzleCLIOptions, 'wrap' | 'eject'>,
): Promise<SwizzleAction> {
  if (options.wrap) {
    return 'wrap';
  }
  if (options.eject) {
    return 'eject';
  }
  return askSwizzleAction(componentConfig);
}

export type ActionParams = {
  siteDir: string;
  themePath: string;
  componentName: string;
  typescript: boolean;
};

export type ActionResult = {
  createdFiles: string[];
};

async function isDir(dirPath: string): Promise<boolean> {
  return (
    (await fs.pathExists(dirPath)) && (await fs.stat(dirPath)).isDirectory()
  );
}

export async function eject({
  siteDir,
  themePath,
  componentName,
  typescript,
}: ActionParams): Promise<ActionResult> {
  const fromPath = path.join(themePath, componentName);
  const isDirectory = await isDir(fromPath);
  const globPattern = isDirectory
    ? // Do we really want to copy all components?
      path.join(fromPath, '**/*')
    : `${fromPath}.*`;

  const globPatternPosix = posixPath(globPattern);

  const filesToCopy = await Globby(globPatternPosix, {
    ignore: _.compact([
      '**/*.{story,stories,test,tests}.{js,jsx,ts,tsx}',
      // When ejecting JS components, we want to avoid emitting TS files
      // In particular the .d.ts files that theme build output contains
      typescript ? null : '**/*.{d.ts,ts,tsx}',
      '**/{__fixtures__,__tests__}/*',
    ]),
  });

  if (filesToCopy.length === 0) {
    // This should never happen
    throw new Error(
      logger.interpolate`No files to copy from path=${fromPath} with glob code=${globPatternPosix}`,
    );
  }

  const toPath = path.join(siteDir, THEME_PATH);

  await fs.ensureDir(toPath);

  const createdFiles = await Promise.all(
    filesToCopy.map(async (sourceFile: string) => {
      const targetFile = path.join(
        toPath,
        path.relative(themePath, sourceFile),
      );
      try {
        const fileContents = await fs.readFile(sourceFile, 'utf-8');
        await fs.outputFile(
          targetFile,
          fileContents.trimStart().replace(/^\/\*.+?\*\/\s*/ms, ''),
        );
      } catch (err) {
        logger.error`Could not copy file from path=${sourceFile} to path=${targetFile}`;
        throw err;
      }
      return targetFile;
    }),
  );
  return {createdFiles};
}

export async function wrap({
  siteDir,
  themePath,
  componentName: themeComponentName,
  typescript,
  importType = 'original',
}: ActionParams & {
  importType?: 'original' | 'init';
}): Promise<ActionResult> {
  const isDirectory = await isDir(path.join(themePath, themeComponentName));

  // Top/Parent/ComponentName => ComponentName
  const componentName = _.last(themeComponentName.split('/'))!;
  const wrapperComponentName = `${componentName}Wrapper`;

  const wrapperFileName = `${themeComponentName}${isDirectory ? '/index' : ''}${
    typescript ? '.tsx' : '.js'
  }`;

  await fs.ensureDir(path.resolve(siteDir, THEME_PATH));

  const toPath = path.resolve(siteDir, THEME_PATH, wrapperFileName);

  const content = typescript
    ? `import React from 'react';
import ${componentName} from '@theme-${importType}/${themeComponentName}';
import type ${componentName}Type from '@theme/${themeComponentName}';
import type {WrapperProps} from '@docusaurus/types';

type Props = WrapperProps<typeof ${componentName}Type>;

export default function ${wrapperComponentName}(props: Props): JSX.Element {
  return (
    <>
      <${componentName} {...props} />
    </>
  );
}
`
    : `import React from 'react';
import ${componentName} from '@theme-${importType}/${themeComponentName}';

export default function ${wrapperComponentName}(props) {
  return (
    <>
      <${componentName} {...props} />
    </>
  );
}
`;

  await fs.outputFile(toPath, content);

  return {createdFiles: [toPath]};
}
