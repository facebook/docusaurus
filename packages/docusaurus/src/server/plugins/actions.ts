/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {generate, posixPath} from '@docusaurus/utils';
import {applyRouteTrailingSlash} from './routeConfig';
import type {
  InitializedPlugin,
  PluginContentLoadedActions,
  PluginRouteContext,
  RouteConfig,
} from '@docusaurus/types';

type PluginActionUtils = {
  getRoutes: () => RouteConfig[];
  getGlobalData: () => unknown;
  getActions: () => PluginContentLoadedActions;
};

// TODO refactor historical action system and make this side-effect-free
//  If the function were pure, we could more easily compare previous/next values
//  on site reloads, and bail-out of the reload process earlier
//  Particularly, createData() modules should rather be declarative
export async function createPluginActionsUtils({
  plugin,
  generatedFilesDir,
  baseUrl,
  trailingSlash,
}: {
  plugin: InitializedPlugin;
  generatedFilesDir: string;
  baseUrl: string;
  trailingSlash: boolean | undefined;
}): Promise<PluginActionUtils> {
  const pluginId = plugin.options.id;
  // Plugins data files are namespaced by pluginName/pluginId

  // TODO Docusaurus v4 breaking change
  //  module aliasing should be automatic
  //  we should never find local absolute FS paths in the codegen registry
  const aliasedSource = (source: string) =>
    `@generated/${posixPath(path.relative(generatedFilesDir, source))}`;

  // TODO use @generated data dir here!
  // The module registry should not contain absolute paths
  const dataDir = path.join(generatedFilesDir, plugin.name, pluginId);

  const pluginRouteContext: PluginRouteContext['plugin'] = {
    name: plugin.name,
    id: pluginId,
  };

  const pluginRouteContextModulePath = path.join(dataDir, `__plugin.json`);

  // TODO not ideal place to generate that file
  // move to codegen step instead!
  await generate(
    '/',
    pluginRouteContextModulePath,
    JSON.stringify(pluginRouteContext, null, 2),
  );

  const routes: RouteConfig[] = [];
  let globalData: unknown;

  const actions: PluginContentLoadedActions = {
    addRoute(initialRouteConfig) {
      // Trailing slash behavior is handled generically for all plugins
      const finalRouteConfig = applyRouteTrailingSlash(initialRouteConfig, {
        baseUrl,
        trailingSlash,
      });
      routes.push({
        ...finalRouteConfig,
        context: {
          ...(finalRouteConfig.context && {data: finalRouteConfig.context}),
          plugin: aliasedSource(pluginRouteContextModulePath),
        },
      });
    },
    async createData(name, data) {
      const modulePath = path.join(dataDir, name);
      const dataString =
        typeof data === 'string' ? data : JSON.stringify(data, null, 2);
      await generate(dataDir, name, dataString);
      return modulePath;
    },
    setGlobalData(data) {
      globalData = data;
    },
  };

  return {
    // Some variables are mutable, so we expose a getter instead of the value
    getRoutes: () => routes,
    getGlobalData: () => globalData,
    getActions: () => actions,
  };
}
