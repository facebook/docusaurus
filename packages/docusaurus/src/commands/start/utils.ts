/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import _ from 'lodash';
import {prepareUrls} from 'react-dev-utils/WebpackDevServerUtils';
import {normalizeUrl} from '@docusaurus/utils';
import logger from '@docusaurus/logger';
import {getHostPort} from '../../server/getHostPort';
import {PerfLogger} from '../../utils';
import {
  loadSite,
  type LoadSiteParams,
  reloadSite,
  reloadSitePlugin,
} from '../../server/site';
import {formatPluginName} from '../../server/plugins/pluginsUtils';
import type {StartCLIOptions} from './start';
import type {LoadedPlugin, RouterType} from '@docusaurus/types';

export type OpenUrlContext = {
  host: string;
  port: number;
  getOpenUrl: ({
    baseUrl,
    router,
  }: {
    baseUrl: string;
    router: RouterType;
  }) => string;
};

export async function createOpenUrlContext({
  cliOptions,
}: {
  cliOptions: StartCLIOptions;
}): Promise<OpenUrlContext> {
  const protocol: string = process.env.HTTPS === 'true' ? 'https' : 'http';

  const {host, port} = await getHostPort(cliOptions);
  if (port === null) {
    return process.exit();
  }

  const getOpenUrl: OpenUrlContext['getOpenUrl'] = ({baseUrl, router}) => {
    const urls = prepareUrls(protocol, host, port);
    return normalizeUrl([
      urls.localUrlForBrowser,
      router === 'hash' ? '/#/' : '',
      baseUrl,
    ]);
  };

  return {host, port, getOpenUrl};
}

type StartParams = {
  siteDirParam: string;
  cliOptions: Partial<StartCLIOptions>;
};

async function createLoadSiteParams({
  siteDirParam,
  cliOptions,
}: StartParams): Promise<LoadSiteParams> {
  const siteDir = await fs.realpath(siteDirParam);
  return {
    siteDir,
    config: cliOptions.config,
    locale: cliOptions.locale,
    localizePath: undefined, // Should this be configurable?
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function createReloadableSite(startParams: StartParams) {
  const openUrlContext = await createOpenUrlContext(startParams);

  const loadSiteParams = await PerfLogger.async('createLoadSiteParams', () =>
    createLoadSiteParams(startParams),
  );

  let site = await PerfLogger.async('Load site', () =>
    loadSite(loadSiteParams),
  );

  const get = () => site;

  const getOpenUrl = () =>
    openUrlContext.getOpenUrl({
      baseUrl: site.props.baseUrl,
      router: site.props.siteConfig.future.experimental_router,
    });

  const printOpenUrlMessage = () => {
    logger.success`Docusaurus website is running at: url=${getOpenUrl()}`;
  };
  printOpenUrlMessage();

  const reloadBase = async () => {
    try {
      const oldSite = site;
      site = await PerfLogger.async('Reload site', () => reloadSite(site));
      if (oldSite.props.baseUrl !== site.props.baseUrl) {
        printOpenUrlMessage();
      }
    } catch (e) {
      logger.error('Site reload failure');
      console.error(e);
    }
  };

  // TODO instead of debouncing we should rather add AbortController support?
  const reload = _.debounce(reloadBase, 500);

  // TODO this could be subject to plugin reloads race conditions
  //  In practice, it is not likely the user will hot reload 2 plugins at once
  //  but we should still support it and probably use a task queuing system
  const reloadPlugin = async (plugin: LoadedPlugin) => {
    try {
      site = await PerfLogger.async(
        `Reload site plugin ${formatPluginName(plugin)}`,
        () => {
          const pluginIdentifier = {name: plugin.name, id: plugin.options.id};
          return reloadSitePlugin(site, pluginIdentifier);
        },
      );
    } catch (e) {
      logger.error(
        `Site plugin reload failure - Plugin ${formatPluginName(plugin)}`,
      );
      console.error(e);
    }
  };

  return {get, getOpenUrl, reload, reloadPlugin, openUrlContext};
}
