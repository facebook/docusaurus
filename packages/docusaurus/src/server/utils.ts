/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import {posixPath, Globby} from '@docusaurus/utils';

// Globby that fix Windows path patterns
// See https://github.com/facebook/docusaurus/pull/4222#issuecomment-795517329
export async function safeGlobby(
  patterns: string[],
  options?: Globby.GlobbyOptions,
): Promise<string[]> {
  // Required for Windows support, as paths using \ should not be used by globby
  // (also using the windows hard drive prefix like c: is not a good idea)
  const globPaths = patterns.map((dirPath) =>
    posixPath(path.relative(process.cwd(), dirPath)),
  );

  return Globby(globPaths, options);
}
