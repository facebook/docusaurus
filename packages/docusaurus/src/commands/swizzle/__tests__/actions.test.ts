/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import fs from 'fs-extra';
import tree from 'tree-node-cli';
import {posixPath} from '@docusaurus/utils';
import {eject, wrap} from '../actions';
import {ThemePath, Components, createTempSiteDir} from './testUtils';
import type {SwizzleAction} from '@docusaurus/types';

// Use relative paths and sort files for tests
function stableCreatedFiles(
  siteThemePath: string,
  createdFiles: string[],
): string[] {
  return createdFiles
    .map((file) => posixPath(path.relative(siteThemePath, file)))
    .sort();
}

describe('eject', () => {
  async function testEject(
    action: SwizzleAction,
    componentName: string,
    {typescript}: {typescript: boolean} = {typescript: true},
  ) {
    const siteDir = await createTempSiteDir();
    const siteThemePath = path.join(siteDir, 'src/theme');
    const result = await eject({
      siteDir,
      componentName,
      themePath: ThemePath,
      typescript,
    });
    return {
      siteDir,
      siteThemePath,
      createdFiles: stableCreatedFiles(siteThemePath, result.createdFiles),
      tree: tree(siteThemePath),
    };
  }

  it(`eject ${Components.FirstLevelComponent}`, async () => {
    const result = await testEject('eject', Components.FirstLevelComponent);
    expect(result.createdFiles).toEqual([
      'FirstLevelComponent.css',
      'FirstLevelComponent.tsx',
    ]);
    expect(result.tree).toMatchInlineSnapshot(`
      "theme
      ├── FirstLevelComponent.css
      └── FirstLevelComponent.tsx"
    `);
  });

  it(`eject ${Components.JsComponent} JS`, async () => {
    const result = await testEject('eject', Components.JsComponent, {
      typescript: false,
    });
    expect(result.createdFiles).toEqual([
      'JsComponent/index.css',
      'JsComponent/index.js',
    ]);
    expect(result.tree).toMatchInlineSnapshot(`
      "theme
      └── JsComponent
          ├── index.css
          └── index.js"
    `);
  });

  it(`eject ${Components.ComponentInSubFolder}`, async () => {
    const result = await testEject('eject', Components.ComponentInSubFolder);
    expect(result.createdFiles).toEqual([
      'ComponentInFolder/ComponentInSubFolder/index.css',
      'ComponentInFolder/ComponentInSubFolder/index.tsx',
      'ComponentInFolder/ComponentInSubFolder/styles.css',
      'ComponentInFolder/ComponentInSubFolder/styles.module.css',
    ]);
    expect(result.tree).toMatchInlineSnapshot(`
      "theme
      └── ComponentInFolder
          └── ComponentInSubFolder
              ├── index.css
              ├── index.tsx
              ├── styles.css
              └── styles.module.css"
    `);
  });

  it(`eject ${Components.Sibling}`, async () => {
    const result = await testEject('eject', Components.Sibling);
    expect(result.createdFiles).toEqual([
      'ComponentInFolder/Sibling.css',
      'ComponentInFolder/Sibling.tsx',
    ]);
    expect(result.tree).toMatchInlineSnapshot(`
      "theme
      └── ComponentInFolder
          ├── Sibling.css
          └── Sibling.tsx"
    `);
  });

  it(`eject ${Components.ComponentInFolder}`, async () => {
    const result = await testEject('eject', Components.ComponentInFolder);
    expect(result.createdFiles).toEqual([
      'ComponentInFolder/ComponentInSubFolder/index.css',
      'ComponentInFolder/ComponentInSubFolder/index.tsx',
      'ComponentInFolder/ComponentInSubFolder/styles.css',
      'ComponentInFolder/ComponentInSubFolder/styles.module.css',
      'ComponentInFolder/Sibling.css',
      'ComponentInFolder/Sibling.tsx',
      'ComponentInFolder/index.css',
      'ComponentInFolder/index.tsx',
    ]);
    expect(result.tree).toMatchInlineSnapshot(`
      "theme
      └── ComponentInFolder
          ├── ComponentInSubFolder
          │   ├── index.css
          │   ├── index.tsx
          │   ├── styles.css
          │   └── styles.module.css
          ├── Sibling.css
          ├── Sibling.tsx
          ├── index.css
          └── index.tsx"
    `);
  });
});

describe('wrap', () => {
  async function testWrap(
    action: SwizzleAction,
    componentName: string,
    {typescript}: {typescript: boolean} = {typescript: false},
  ) {
    const siteDir = await createTempSiteDir();
    const siteThemePath = path.join(siteDir, 'src/theme');
    const result = await wrap({
      siteDir,
      componentName,
      themePath: ThemePath,
      typescript,
    });
    return {
      siteDir,
      siteThemePath,
      createdFiles: stableCreatedFiles(siteThemePath, result.createdFiles),
      firstFileContent: () => fs.readFile(result.createdFiles[0]!, 'utf8'),
      tree: tree(siteThemePath),
    };
  }

  describe('JavaScript', () => {
    async function doWrap(componentName: string) {
      return testWrap('wrap', componentName, {
        typescript: false,
      });
    }

    it(`wrap ${Components.FirstLevelComponent}`, async () => {
      const result = await doWrap(Components.FirstLevelComponent);
      expect(result.createdFiles).toEqual(['FirstLevelComponent.js']);
      expect(result.tree).toMatchInlineSnapshot(`
              "theme
              └── FirstLevelComponent.js"
          `);
      await expect(result.firstFileContent()).resolves.toMatchSnapshot();
    });

    it(`wrap ${Components.ComponentInSubFolder}`, async () => {
      const result = await doWrap(Components.ComponentInSubFolder);
      expect(result.createdFiles).toEqual([
        'ComponentInFolder/ComponentInSubFolder/index.js',
      ]);
      expect(result.tree).toMatchInlineSnapshot(`
              "theme
              └── ComponentInFolder
                  └── ComponentInSubFolder
                      └── index.js"
          `);
      await expect(result.firstFileContent()).resolves.toMatchSnapshot();
    });

    it(`wrap ${Components.ComponentInFolder}`, async () => {
      const result = await doWrap(Components.ComponentInFolder);
      expect(result.createdFiles).toEqual(['ComponentInFolder/index.js']);
      expect(result.tree).toMatchInlineSnapshot(`
              "theme
              └── ComponentInFolder
                  └── index.js"
          `);
      await expect(result.firstFileContent()).resolves.toMatchSnapshot();
    });
  });

  describe('TypeScript', () => {
    async function doWrap(componentName: string) {
      return testWrap('wrap', componentName, {
        typescript: true,
      });
    }

    it(`wrap ${Components.FirstLevelComponent}`, async () => {
      const result = await doWrap(Components.FirstLevelComponent);
      expect(result.createdFiles).toEqual(['FirstLevelComponent.tsx']);
      expect(result.tree).toMatchInlineSnapshot(`
              "theme
              └── FirstLevelComponent.tsx"
          `);
      await expect(result.firstFileContent()).resolves.toMatchSnapshot();
    });

    it(`wrap ${Components.ComponentInSubFolder}`, async () => {
      const result = await doWrap(Components.ComponentInSubFolder);
      expect(result.createdFiles).toEqual([
        'ComponentInFolder/ComponentInSubFolder/index.tsx',
      ]);
      expect(result.tree).toMatchInlineSnapshot(`
              "theme
              └── ComponentInFolder
                  └── ComponentInSubFolder
                      └── index.tsx"
          `);
      await expect(result.firstFileContent()).resolves.toMatchSnapshot();
    });

    it(`wrap ${Components.ComponentInFolder}`, async () => {
      const result = await doWrap(Components.ComponentInFolder);
      expect(result.createdFiles).toEqual(['ComponentInFolder/index.tsx']);
      expect(result.tree).toMatchInlineSnapshot(`
              "theme
              └── ComponentInFolder
                  └── index.tsx"
          `);
      await expect(result.firstFileContent()).resolves.toMatchSnapshot();
    });
  });
});
