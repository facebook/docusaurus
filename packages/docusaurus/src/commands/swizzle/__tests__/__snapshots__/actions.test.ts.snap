// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrap JavaScript wrap ComponentInFolder 2`] = `
"import React from 'react';
import ComponentInFolder from '@theme-original/ComponentInFolder';

export default function ComponentInFolderWrapper(props) {
  return (
    <>
      <ComponentInFolder {...props} />
    </>
  );
}
"
`;

exports[`wrap JavaScript wrap ComponentInFolder/ComponentInSubFolder 2`] = `
"import React from 'react';
import ComponentInSubFolder from '@theme-original/ComponentInFolder/ComponentInSubFolder';

export default function ComponentInSubFolderWrapper(props) {
  return (
    <>
      <ComponentInSubFolder {...props} />
    </>
  );
}
"
`;

exports[`wrap JavaScript wrap FirstLevelComponent 2`] = `
"import React from 'react';
import FirstLevelComponent from '@theme-original/FirstLevelComponent';

export default function FirstLevelComponentWrapper(props) {
  return (
    <>
      <FirstLevelComponent {...props} />
    </>
  );
}
"
`;

exports[`wrap TypeScript wrap ComponentInFolder 2`] = `
"import React from 'react';
import ComponentInFolder from '@theme-original/ComponentInFolder';
import type ComponentInFolderType from '@theme/ComponentInFolder';
import type {WrapperProps} from '@docusaurus/types';

type Props = WrapperProps<typeof ComponentInFolderType>;

export default function ComponentInFolderWrapper(props: Props): JSX.Element {
  return (
    <>
      <ComponentInFolder {...props} />
    </>
  );
}
"
`;

exports[`wrap TypeScript wrap ComponentInFolder/ComponentInSubFolder 2`] = `
"import React from 'react';
import ComponentInSubFolder from '@theme-original/ComponentInFolder/ComponentInSubFolder';
import type ComponentInSubFolderType from '@theme/ComponentInFolder/ComponentInSubFolder';
import type {WrapperProps} from '@docusaurus/types';

type Props = WrapperProps<typeof ComponentInSubFolderType>;

export default function ComponentInSubFolderWrapper(props: Props): JSX.Element {
  return (
    <>
      <ComponentInSubFolder {...props} />
    </>
  );
}
"
`;

exports[`wrap TypeScript wrap FirstLevelComponent 2`] = `
"import React from 'react';
import FirstLevelComponent from '@theme-original/FirstLevelComponent';
import type FirstLevelComponentType from '@theme/FirstLevelComponent';
import type {WrapperProps} from '@docusaurus/types';

type Props = WrapperProps<typeof FirstLevelComponentType>;

export default function FirstLevelComponentWrapper(props: Props): JSX.Element {
  return (
    <>
      <FirstLevelComponent {...props} />
    </>
  );
}
"
`;
